Collector/ProfileClient.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+namespace Http\HttplugBundle\Collector;
+
+use Http\Client\Common\FlexibleHttpClient;
+use Http\Client\HttpAsyncClient;
+use Http\Client\HttpClient;
+use Psr\Http\Message\RequestInterface;
+
+/**
+ * The ProfileClient decorates any client that implement both HttpClient and HttpAsyncClient interfaces to gather target
+ * url and response status code.
+ *
+ * @author Fabien Bourigault <bourigaultfabien@gmail.com>
+ *
+ * @internal
+ */
+class ProfileClient implements HttpClient, HttpAsyncClient
+{
+    /**
+     * @var HttpClient|HttpAsyncClient
+     */
+    private $client;
+
+    /**
+     * @var Collector
+     */
+    private $collector;
+
+    /**
+     * @param HttpClient|HttpAsyncClient $client    The client to profile. Client must implement both HttpClient and
```

Comment:
To be honest, I don't understand what inversion of control has to do with this, but I trust you have you (maybe fairly complicated reasons). Plus, I wasn't aware there were other capabilities, and if an interface is introduced for every combination, this could get quite messy… the real thing to fix here might be php itself , to add the possibility to type hint against several interfaces :P